map.go

```diff
@@ -464,11 +467,17 @@ func (m *Map[K, V]) wasNeverFull(i uintptr) bool {
 	indexBefore := (i - groupSize) & m.capacity
 	emptyAfter := m.ctrls.At(i).matchEmpty()
 	emptyBefore := m.ctrls.At(indexBefore).matchEmpty()
+	if debug {
+		fmt.Printf("wasNeverFull: before=%d/%s/%d after=%d/%s/%d\n",
+			indexBefore, emptyBefore, bits.LeadingZeros64(uint64(emptyBefore))>>3,
+			i, emptyAfter, bits.TrailingZeros64(uint64(emptyAfter))>>3)
+	}
 	// We count how many consecutive non empties we have to the right and to
 	// the left of i. If the sum is >= groupSize then there is at least one
 	// probe window that might have seen a full group.
 	if emptyBefore != 0 && emptyAfter != 0 &&
-		(emptyBefore.count()+emptyAfter.count()) < groupSize {
+		((bits.TrailingZeros64(uint64(emptyAfter))>>3)+
```

Comment:
I've expanded the comment and added a diagram to make it clearer what is going on. PTAL. And definitely ping me if there are other bits of subtle code. I have the understanding of the Abseil Swiss Table implementation all in my. head right now, but it will disappear.

Reply:
Thanks, this helps a lot. I don't know why I was confused before.

I have a few places where we could use some more comments but they're not strictly related to this change. I think I'll give it a go at writing them and open a PR.

One thing that I found confusing is how we do `ctrls.At(i)` which gives us a pointer to an uint8, and then we call something like `matchEmpty` on `*ctrl` which casts the receiver to a `*uint64` (this last part especially is far beyond "normal usage").  I think instead we should have a `ctrls` type which stores a pointer like `unsafeSlice` but gives you accessors for both the `*ctrl` and `bitset`. Methods like `matchEmpty` would just be methods on `bitset`.

Also `bitset` should maybe be called `group` or `ctrlGroup`

If you think these are good ideas I can take a stab at it.

Reply:
> If you think these are good ideas I can take a stab at it.

I've gotten used to the existing setup, but I'm not attached to it. Willing to see what you're thinking and if it makes the code clearer.